# Fundamentals of Data Analysis Tasks Notebook

Author - Sean Humphreys

---

### Contents

1. [Task 1 - Collatz Conjecture](#task-1---collatz-conjecture)
    1. [Step 1 - Test the Collatz Conjecture](#step-1---test-collatz-conjecture)
    2. [Step 2 - Test a Range of Input Values](#step-2---test-a-range-of-input-values)
    3. [Step 3 - Tell the User the Result](#step-3---tell-the-user-the-result)
    4. [Step 4 - Make the Script Re-Usable](#step-4---make-the-script-re-useable)
    5. [References](#task-1-references)
    6. [Background Reading](#task-1-background-reading)


2. [Task 2 - Penguins Dataset](#task-2---penguins-dataset)
    1. [Dataset Variables](#dataset-variables)
        1. [Variable Classification](#variable-classification)
        2. [Missing Values](#missing-values)
    2. [Variable Selection](#variable-selection)
    2. [Task 2 References](#task-2-references)
    3. [Task 2 Background Reading](#task-2-background-reading)

## Task 1 - Collatz Conjecture <a id="task-1---collatz-conjecture"></a>
----

Task 1 - *Verify, using Python, that the conjecture is true for the first 10,000 positive integers.*

Often thought of as the simplest impossible problem the Collatz Conjecture asks whether repeating two simple arithmetic operations will eventually transform every positive integer into 1.

$f(n) = \begin{cases} n/2 &\text{if } n = \text{even} \\
(3n +1)/2 & \text{if } n = \text{odd} \end{cases}$


**Assumptions**
- If the Collatz conjecture for a positive integer holds true all numbers will end in an infinite loop of 4, 2, 1 - this includes the numbers 1 and 2.
- The script should indicate to the end user if the Collatz Conjecture holds true for the desired range of positive integers.
- If the Collatz conjecture does not hold true for any values, the script should return the values that do not hold true.
- For usability purposes the script should be able to handle any range of input values not just the first 10,000 positive integers.
- The script will not run if negative integers, floating point decimals or strings are entered as arguments.
- If these values are entered as arguments the script will give clear instruction to the end user of the correct format for arguments to be entered.

### Step 1 - Test Collatz Conjecture <a id="step-1---test-collatz-conjecture"></a>

Define a function to test the Collatz conjecture. 

This function is based on a code snippet accessible [here](https://www.educative.io/answers/how-to-generate-the-collatz-sequence-in-python) (last accessed 06/10/2023). 

The purpose of the *collatz(number)* function is to return the Collatz sequence, in a list, for a given positive integer. If the Collatz Conjecture does not hold true for that number a list containing the given number is returned. To achieve this the Collatz sequence for the given number is tested against a list that contains the numbers 4, 2, 1 - as per the assumption that every Collatz sequence must end in a 4, 2, 1 loop.

The code snippet upon which the function is based calculates the the Collatz sequence for the number 1 as [1] and number 2 as [2, 1]. This creates an issue in so far as the sequence for the numbers 1 and 2 cannot be tested against the [4, 2, 1] loop. To allow for this, an *if* and *elif* statement are used to return the Collatz sequence for the numbers 1 and 2 that include the 4, 2, 1 loop. All other values passed as an argument to the function are divided by 2 if even or if odd they multiplied by 3 and 1 is added to this number. The returned value is appended to a list. A *while* loop is employed to continue this until the output of these mathematical operations reaches 1.

The *collatz_sequence_list* generated by the while loop is tested for Collatz compliance by comparing the last three numbers in this list against the *loop_list*. If false an *if* statement returns the *wrong_list*. If true, the function returns the *collatz_sequence_list*.

In [1]:
def collatz(number):
    collatz_sequence_list = []
    wrong_list = []
    loop_list = [4, 2, 1]
    if number == 1:
        # to test 1 the 4,2,1 loop needs to be included to test against
        collatz_sequence_list = [1, 4, 2, 1]
    elif number == 2:
        # to test 2 the 4,2,1 loop needs to be included to test against
        collatz_sequence_list = [2, 4, 2, 1]
    else:
        collatz_sequence_list.append(number)
        while (number != 1):
            if (number % 2 == 0):
                number = number//2
                collatz_sequence_list.append(number)
            else:
                number = number*3+1
                collatz_sequence_list.append(number)
    if loop_list != collatz_sequence_list[-3:]:
        wrong_list.append(collatz_sequence_list[0])
        return wrong_list
    else:
        return collatz_sequence_list

The output of the `collatz(number)` function is tested below. For the purpose of this test, number = 57.

In [2]:
print(collatz(57))

[57, 172, 86, 43, 130, 65, 196, 98, 49, 148, 74, 37, 112, 56, 28, 14, 7, 22, 11, 34, 17, 52, 26, 13, 40, 20, 10, 5, 16, 8, 4, 2, 1]


### Step 2 - Test a Range of Input Values <a id="step-2---test-a-range-of-input-values"></a>

Define a function to test a range of values. 

Using a *for* loop, every value in a given range of values can be inputted into the into the `collatz(number)` function. There are two possible outcomes that must be accounted for:
1. The number does not satisfy the Collatz conjecture. In this outcome a list 1 number long is returned.
2. The number does satisfy the Collatz conjecture. In this outcome a list longer than 1 number is returned.

Each of these outcomes is handled separately. Two functions are defined to achieve this.

1. The `wrong_list(lower, upper)` function uses a *for* loop to check the length of the list returned from the `collatz(number)` function for every value in the desired range. If the length of the returned list is 1, i.e. the Collatz conjecture does not hold true for that number, that number is appended to a list and returned by the function on completion of the *for* loop.

In [3]:
def wrong_list(lower, upper):
    wrong_list = []
    for i in range(lower, upper):
        if len(collatz(i)) == 1:
            wrong_list.append(i)
    return (wrong_list)

2. The `right_list(lower, upper)` function uses a *for* loop to check the length of the list returned from the `collatz(number)` function for every value in the desired range. If the length of the returned list is greater than 1, i.e. the Collatz conjecture is true for that number, that number is appended to a list and the returned by the function on completion of the *for* loop.

In [4]:
def right_list(lower, upper):
    right_list = []
    for i in range(lower, upper):
        if len(collatz(i)) > 1:
            right_list.append(i)
    return (right_list)

### Step 3 - Tell the User the Result <a id="step-3---tell-the-user-the-result"></a>

Define a function to tell the user the vales for which the Collatz conjecture does or does not hold true for in a given range.

The input arguments for this function are the lower and upper values for the range of values to be tested.

4 variables are defined:
1. wrong_list_var = a list containing the output of the `wrong_list(lower, upper)` function for the range of values given as arguments 
2. right_list_var = a list containing the output of the `right_list(lower, upper)` function for the range of values given as arguments 
3. string = string to be printed
4. result = list comprehension to print a list of lists without brackets

If the length of the list returned from `wrong_list(lowest_number, highest_number)` is greater than 1 the concatenated *string* and *result* variables will be printed. Else, if, the length of the list returned from `right_list(lowest_number, highest_number)` is greater than 1 a formatted string that confirms the Collatz conjecture for the given range of values to be tested is printed.

In [5]:
def result(lowest_number, highest_number):
    wrong_list_var = wrong_list(lowest_number, highest_number)
    right_list_var = right_list(lowest_number, highest_number)
    string = 'The numbers that Collatz Conjecture does not hold true for are: '
    result = ', '.join(str(item) for item in wrong_list_var)
    if len(wrong_list_var) > 0:
        print(string + result)
    elif len(right_list_var) > 0:
        print(
            f'The Collatz Conjecture holds true for the numbers from {lowest_number} up to and including {highest_number-1}.')

The `result(lowest_number, highest number)` function output is demonstrated below for the first 10,000 positive integers.

In [6]:
result(1,10001)

The Collatz Conjecture holds true for the numbers from 1 up to and including 10000.


### Step 4 - Make the Script Re-useable <a id="step-4---make-the-script-re-useable"></a>

Importing the sys library will allow for better re-useability of the script. The script can be run from the command line with the range of values to be tested passed as arguments from the command line.

In [7]:
import sys

The following variables are defined and casted to integers:
1. low_number = sys.argv[1]. sys.argv[1] is the first argument passed from the command line. This should be the start of the range to test. The risk is that the user enters a negative value, a floating point integer or a string.  
2. high_number = sys.argv[2] + 1. sys.argv[2] is the second argument passed from the command line. This should be the end of the range to be test. 1 is added to this value. The risk is that the user enters a negative value, a floating point integer, a string or a value that is lower than sys.argv[1].

To mitigate against the risks outlined above a combination of error handling, *if*, *elif* and *else* statements are used. The code contained in the *try* block will run unless a *ValueError* or an *IndexError* are caught. If this happens instructions are printed advising the user of the corrective action to take on re-run of the script. If no error is caught the `result(low_number, high_number)` function will test the range of values given as command line arguments only if the *low_number* value is a positive integer that is less than the *high_number* variable.

In [8]:
try:
    low_number = int(sys.argv[1]) # cast to int
    high_number = int(sys.argv[2]) + 1 # need to add 1 for desired range
    if low_number > 0 and low_number < high_number:
        result(low_number, high_number)
    elif low_number > 0 and high_number < low_number:
        print(f'{high_number} is lower than {low_number}. The script requires that the second argument is greater than'
              f' the first. Please run the script again with the correct parameters.')
    else:
        print(f'{low_number} is not a positive integer. Please run the script again and enter a positive integer as '
              f'an argument.')
except ValueError: # error handling
    print(f'Please enter a positive integer as an argument.')
except IndexError: # error handling
    print('Please enter two command line arguments. The arguments must be positive integers with the first number '
          f'less than the second number.')

Please enter a positive integer as an argument.


In the execution of the code above error handling is demonstrated. As the *low_number* variable is not a positive integer the *ValueError* code block is executed.  

### Task 1 References <a id="task-1-references"></a>

Educative: Interactive Courses for Software Developers. (n.d.). How to generate the Collatz sequence in Python. [online] Available at: https://www.educative.io/answers/how-to-generate-the-collatz-sequence-in-python. [accessed 06 Oct. 2023].


### Task 1 Background Reading <a id="task-1-background-reading"></a>

bobbyhadz.com. (n.d.). Print a List without the Commas and Brackets in Python | bobbyhadz. [online] Available at: https://bobbyhadz.com/blog/python-print-list-without-commas-and-brackets [Accessed 6 Oct. 2023].

Chaudhuri, D.A.K. (2020). Collatz Conjecture-the simplest impossible problem. [online] Cooking Cosmos. Available at: https://asischaudhuri.wordpress.com/2020/11/09/collatz-conjecture/ [Accessed 6 Oct. 2023].

Machine Intelligence ? (2008). LaTeX – Multiline equations, systems and matrices. [online] Available at: https://kogler.wordpress.com/2008/03/21/latex-multiline-equations-systems-and-matrices/ [Accessed 6 Oct. 2023].

www.w3schools.com. (n.d.). Python Try Except. [online] Available at: https://www.w3schools.com/python/python_try_except.asp. [Accessed 6 Oct. 2023].

---

## Task 2 - Penguins Dataset <a id="task-2---penguins-dataset"></a>
---

Task 2 - *Give an overview of the famous penguins data set, explaining the types of variables it contains. Suggest the types of variables that should be used to model them in Python, explaining your rationale.*

The **Penguins** dataset contains size measurements for three penguin species observed on three islands in the Palmer Archipelago, Antarctica. The data was as part of the Palmer Station, Antarctica, Long-Term Ecological Research program and is also known as the Palmer Penguins dataset. The research paper on the Palmer Penguins dataset by Gorman, Williams and Fraser (2014) is available [here](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0090081#s2).

![image 1](images/illustrations/penguins.png "Palmer Penguins")

Mednick, 2021

### Dataset Variables <a id="dataset-variables"></a>

Using *Python* the penguins dataset is imported. To do this the *Pandas* and *Seaborn* libraries are imported.

[Pandas](https://pandas.pydata.org/) is an open source software library used in data analytics that allows data analysis and manipulation. Pandas is built on top of the *Python* programming language. A Pandas DataFrame is a dictionary like container for Series objects. A DataFrame is the primary Pandas data structure. Using Pandas it is possible to define variables to put the Iris dataset into Pandas DataFrames.

[Seaborn](https://seaborn.pydata.org/) is an open-source software package used for visualisation of statistical data. It is based on on *matplotlib*.

In [9]:
# import the required software libraries
import pandas as pd
import seaborn as sns
penguins = sns.load_dataset('penguins')

Conveniently, the Palmer Penguins dataset is available in *Seaborn*. Once loaded the `info()` function can be used to inspect the dataset.

In [10]:
penguins = sns.load_dataset('penguins')

# check penguins is a pandas DataFrame
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


The summary above demonstrates the class as a *Pandas* DataFrame. There are $344$ rows and seven columns in the dataset. The data is a combination of objects (columns $0$, $1$ & $6$) and $64$-bit floating point numbers (columns $2$, $3$, $4$ & $5$). $64$-bit datatypes allow for a hgh degree of precision. However, use of $64$-bit precision comes at a cost of computational overhead as more memory is required to complete calculations on these numbers due to their size.

#### Variable Classification <a id="variable-classification"></a>

Variable classification determines the type of analysis that can be carried out on that variable. The graphic below summarises the types of data and how they can be analysed. 

![Data Summary](images/illustrations/data_categories_18102023_sh.png "Data Summary Graphic")

Summarised from Sharma, 2020

The first $5$ rows of the DataFrame can be viewed using the `head()` function. This is supports data variable classification.


In [11]:
# print the first five rows of the dataset
penguins.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

Levels of measurement tell how precisely variables are recorded (Bhandari, 2020) and are summarised as follows:
| Level             | Description                                                                   |
| :-----------      | :----------                                                                   |
| Nominal           | The data can only be categorized                                              | 
| Ordinal           | The data can be categorized and ranked                                        | 
| Interval          | The data can be categorized, ranked, and evenly spaced                        | 
| Ratio             | The data can be categorized, ranked, evenly spaced, and has a natural zero    | 

Applying variable classification and level of measurement the Penguins dataset can be summarised as follows:
| Variable          | Classification    | Type          | Python Dtype  | Unit of Measure   | Level of Measure  |
| :----------       | :----------       | :----------   | :----------   | :---------        | :---------        |
| species           | Categorical       | Nominal       | Object        | N/A               | Nominal           |
| island            | Categorical       | Nominal       | Object        | N/A               | Nominal           |
| bill_length_mm    | Numerical         | Continuous    | float$64$     | mm                | Ratio             |
| bill_depth_mm     | Numerical         | Continuous    | float$64$     | mm                | Ratio             |
| flipper_length_mm | Numerical         | Continuous    | float$64$     | mm                | Ratio             |
| body_mass_g       | Numerical         | Continuous    | float$64$     | g                 | Ratio             |
| sex               | Categorical       | Nominal       | Object        | N/A               | Nominal           |

#### Missing Values <a id="missing-values"></a>

The presence of null values in the dataset is confirmed in the "Non-Null Count" column. Columns indexed 2-6 inclusive all have null values as the *Non-Null Count* for these columns is less than $344$. The presence of missing values in any dataset is an issue in so far as it can create a lack of relatability or trustworthiness in the dataset (Kumar, 2021). Consequently, if a strategy to manage missing values during model development is not employed insight could be lost leading to incorrect results (Tamboli, 2021). To summarise the missing values the `isnull()` function chained with the `sum()` function.

In [12]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

### Variable Selection <a id="variable-selection"></a>
The *species*, *island* and *sex* variables are categorical variable types that are nominal. The mathematical operations that can be performed on these variables for analysis are limited to counts of objects. Therefore these variables are not suitable to use as variables for modelling in python.

*Bill length*, *bill depth*, *flipper length* and *body mass* are all numerical variables that are continuous. The level of measure is ratio as the units of each of these variables have an absolute zero e.g. mass or length cannot have a negative value. These variables are classed as continuous as there is an infinite number of values between each number on their scale e.g. there are an infinite number of values between $1$mm and $2$mm and there are an infinite number of values between $1$g and $2$g. As per Sharma (2020), continuous data mathematical operations can include fractions and decimals. All of the numerical data in the Penguins dataset are $64$-bit floating point numbers. This is an issue due to the limitations in the way in which Python performs floating point arithmetic (docs.python.org, n.d.). Floating point numbers are stored in memory as base $2$ binary fractions. Most decimal fractions cannot be represented exactly as binary fractions. Thus decimal floating-point numbers are only approximated by the binary floating-point numbers actually stored in the machine. This has the potential to cause inaccuracies when arithmetic is performed on floating point numbers in Python. In the *Penguins* dataset *Bill length*, *bill depth*, *flipper length* and *body mass* stored in Python as $64$bit floating point numbers. This allows for a high degree of precision and will somewhat mitigate the issue outlined above, albeit at the cost of a computational overhead due to the larger size of this datatype.

Within the *Penguins* dataset *Bill length*, *bill depth*, *flipper length* and *body mass* should be used to model the dataset because:
1. These variables are classified as continuous variables.
2. The level of measure for these variables is ratio.
3. Complex mathematical arithmetic can be performed on these variable types (addition, subtraction, multiplication and division).
4. Mode, median, arithmetic mean and geometric mean can be calculated on these data types.
5. Range, interquartile range, standard deviation, variance & relative standard deviation can be calculated on these variables.
6. Measures of correlation between these variables can be quantified.
7. These variables are stored as $64$-bit floating point numbers which allows for a high degree of precision which allows for more accurate models.
8. Strategies can be adopted to manage the impact of missing values in the dataset when performing models so that the risk of incorrect results, when modelling, is minimised.

### Task 2 References <a id="task-2-references"></a>

Bhandari, P. (2020). Levels of Measurement | Nominal, Ordinal, Interval and Ratio. [online] Scribbr. Available at: https://www.scribbr.com/statistics/levels-of-measurement/. [Accessed 18 Oct. 2023].

docs.python.org. (n.d.). 15. Floating Point Arithmetic: Issues and Limitations — Python 3.10.1 documentation. [online] Available at: https://docs.python.org/3/tutorial/floatingpoint.html. [Accessed 18 Oct. 2023].

Gorman, K.B., Williams, T.D. and Fraser, W.R. (2014). Ecological Sexual Dimorphism and Environmental Variability within a Community of Antarctic Penguins (Genus Pygoscelis). PLoS ONE, 9(3), p.e90081. doi:https://doi.org/10.1371/journal.pone.0090081. [Accessed 16 Oct. 2023]

Kumar, S. (2021). Deep Dive Analysis of Missing Values in Dataset. [online] Medium. Available at: https://towardsdatascience.com/deep-dive-analysis-of-missing-values-in-dataset-b387d9de6d4b. [Accessed 18 Oct. 2023].

PhD, G.M. (2021). Palmer penguins. [online] Gabe Mednick. Available at: https://www.gabemednick.com/post/penguin/ [Accessed 16 Oct. 2023].

Sharma, R. (2020). 4 Types of Data: Nominal, Ordinal, Discrete, Continuous. [online] upGrad blog. Available at: https://www.upgrad.com/blog/types-of-data/. [Accessed 18 Oct. 2023].

Tamboli, N. (2021). Tackling Missing Value in Dataset. [online] Analytics Vidhya. Available at: https://www.analyticsvidhya.com/blog/2021/10/handling-missing-value/. [Accessed 18 )ct. 2023].

### Task 2 Background Reading <a id="task-2-background-reading"></a>

archive-beta.ics.uci.edu. (n.d.). UCI Machine Learning Repository. [online] Available at: https://archive-beta.ics.uci.edu/dataset/690/palmer+penguins-3 [Accessed 16 Oct. 2023].

allisonhorst.github.io. (n.d.). Introduction to palmerpenguins. [online] Available at: https://allisonhorst.github.io/palmerpenguins/articles/intro.html. [Accessed 16 Oct. 2023].

Pandas (2018). Python Data Analysis Library — pandas: Python Data Analysis Library. [online] Pydata.org. Available at: https://pandas.pydata.org/. [Accessed 13 Oct. 2023].

seaborn (2012). seaborn: statistical data visualization — seaborn 0.9.0 documentation. [online] Pydata.org. Available at: https://seaborn.pydata.org/. [Accessed 17 Oct. 2023].

Stack Overflow. (n.d.). The real difference between float32 and float64. [online] Available at: https://stackoverflow.com/questions/43440821/the-real-difference-between-float32-and-float64 [Accessed 17 Oct. 2023].

Tsai, E. (2022). Build a Python project using Pandas and Seaborn | Eugene Tsai. [online] Medium. Available at: https://medium.com/@marvelouskgc/build-a-penguin-project-using-pandas-and-seaborn-eugene-tsai-e4b7e0b499ea [Accessed 16 Oct. 2023].

www.w3schools.com. (n.d.). Pandas DataFrame head() Method. [online] Available at: https://www.w3schools.com/python/pandas/ref_df_head.asp. [Accessed 16 Oct. 2023].

---